crowbar_framework/app/models/service_object.rb

```diff
@@ -1170,6 +1174,10 @@ def apply_role(role, inst, in_queue, bootstrap = false)
         # If new_nodes is empty, we are just removing the proposal.
         unless new_nodes.empty?
           new_nodes.each do |node_name|
+            # skip adding nodes to the batch unless they are really in the list to be deployed
+            # do it before the Node load to avoid doing the call if the node is not there,
+            # as we dont want to spend cycles doing extra calls not needed
```

Comment:
Ah, actually, I'd request a comment like: "This also means we don't ensure that the node has all required roles (through the use of pending_node_actions); it's a reasonable trade-off in the context of this specific optimization, as the nodes should have all roles already (unless there's a bug or the customer broke things manually).".

Okay, it's too long, but you get the idea ;-)